core/arch/arm/plat-rcar/hw_rng.c

```diff
@@ -22,26 +22,32 @@ static unsigned int spin_lock __nex_data = SPINLOCK_UNLOCK;
  */
 static uint8_t rng_cache[PLAT_RND_VECTOR_SZ] __nex_bss
 					__aligned(RCAR_CACHE_LINE_SZ);
-static uint8_t rng_cache_pos __nex_data = PLAT_RND_VECTOR_SZ;
+static uint8_t rng_cache_pos __nex_data;
 
-uint8_t hw_get_random_byte(void)
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
 {
 	uint32_t exceptions = cpu_spin_lock_xsave(&spin_lock);
+	uint8_t *buffer = (uint8_t *)buf;
+	size_t buffer_pos = 0;
 	uint8_t ret_val = 0;
 
-	assert(rng_cache_pos <= PLAT_RND_VECTOR_SZ);
-
-	if (rng_cache_pos == PLAT_RND_VECTOR_SZ) {
-		uint32_t ret = plat_rom_getrndvector(rng_cache, scratch_buf,
-						     sizeof(scratch_buf));
-
-		if (ret != 0)
-			panic("ROM_GetRndVector() returned error!");
-
-		rng_cache_pos = 0;
+	assert(rng_cache_pos < PLAT_RND_VECTOR_SZ);
+
+	while (buffer_pos < len) {
```

Comment:
is it ok to hold a spinlock  during this loop?

Maybe use
```patch
-uint8_t hw_get_random_byte(void)
+static uint8_t __hw_get_random_byte(void)
 (...)
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+   while (len--)
+       *((uint8_t *)buf)++ = __hw_get_random_byte();
+}

Reply:
Not sure why it wouldn't be okay, unless you are suggesting we use something lighter like a regular `mutex_lock`? We need to serialize inside these calls anyway since they update a shared buffer/index.

Reply:
For these drivers, the existing spinlock already protects IOMEM accesses. I don't think all those platform are okay to hold a spinlock and more specifically mask all interrupts by the time a caller gets all its expected random bytes.
All in one, I think the spinlock_xsave/xrestore are ok but if locked/unlocked inside the loop, for each chunk of bytes read.